File: easy/15. Product Sum/product_sum.go
package main

type SpecialArray []interface{}

func _sum(array []interface{}, depth int) int {
	var sum int
	for _, v := range array {
		switch v.(type) {
		case SpecialArray:
			sum += _sum(v.(SpecialArray), depth+1)
		default:
			sum += v.(int)
		}
	}
	return sum * depth
}

// Tip: Each element of `array` can either be cast
// to `SpecialArray` or `int`.
func ProductSum(array []interface{}) int {
	return _sum(array, 1)
}

File: easy/15. Product Sum/product_sum.cpp
#include <any>
#include <vector>

// Tip: You can use el.type() == typeid(vector<any>) to check whether an item is
// a list or an integer.
// If you know an element of the array is vector<any> you can cast it using:
//     any_cast<vector<any>>(element)
// If you know an element of the array is an int you can cast it using:
//     any_cast<int>(element)
int productSum(std::vector<any> array, int depth = 1) {
    int sum = 0;
    for (auto& el : array) {
        if (el.type() == typeid(vector<any>)) {
            sum += productSum(any_cast<vector<any>>(el), depth + 1);
        } else {
            sum += any_cast<int>(el);
        }
    }
    return sum * depth;
  }
